### PR TITLE
CLI helper: Wrap

### PR DESCRIPTION
**Pull request by grod220**
The second CLI helper (after https://github.com/solana-program/token-wrap/pull/33) for the `Wrap` instruction. Note, it only supports single signer at the moment. 

## Walkthrough for testing locally:

1. Build program
```bash
cargo build-sbf # generates .so file
cargo build # builds binary
```

2. Start solana test validator
```bash
solana-test-validator \
  --bpf-program TwRapQCDhWkZRrDaHfZGuHxkZ91gHDRkyuzNqeU5MgR target/deploy/spl_token_wrap.so \
  --reset
```

3. Create a new mint & token account
```bash
spl-token create-token
# mint addr -> G85DDQCoHchhDFWtUFT2MDkfUPYQt8BvViFyjGv167xb
spl-token create-account G85DDQCoHchhDFWtUFT2MDkfUPYQt8BvViFyjGv167xb 
# token account -> EYUCh7xP4sxnNBr1VFcGXfHm1piEA6ERci5prNykSV4G
spl-token mint G85DDQCoHchhDFWtUFT2MDkfUPYQt8BvViFyjGv167xb 100000
```

4. Run CreateMint `token-wrap` command
```bash
cargo run --bin spl-token-wrap create-mint \
    G85DDQCoHchhDFWtUFT2MDkfUPYQt8BvViFyjGv167xb \  # unwrapped_mint
    TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA \    # unwrapped_token_program
    TokenzQdBNbLqP5VEhdkAS6EPFLC1PHnBqCXEpPxuEb     # wrapped_token_program

# Keep note of console output
# Unwrapped mint address: 8eZ2rk3GcKopnKrAmVibV7zLW1sszNo2GdbhtxRhrZRi
# Wrapped mint address: 3xnGREX1Mi7zcuXJAYoDGk9xjpugZnfRhAKLPHdfEPPq
# Wrapped mint authority: A2LJVqpdm8kQUpw9x6PBZNVKzG9KesxF7gpbbvHxe8m
```

5. Create an escrow token account and set its owner to the mint_authority
```bash
spl-token create-account 8eZ2rk3GcKopnKrAmVibV7zLW1sszNo2GdbhtxRhrZRi --owner A2LJVqpdm8kQUpw9x6PBZNVKzG9KesxF7gpbbvHxe8m --fee-payer /Users/gabe/.config/solana/id.json
```

6. Create a recipient token account of that new wrapped mint
```bash
spl-token create-account 3xnGREX1Mi7zcuXJAYoDGk9xjpugZnfRhAKLPHdfEPPq
```

7. Issue Wrap command
```bash
spl-token-wrap wrap <UNWRAPPED_TOKEN_PROGRAM> <UNWRAPPED_MINT> <UNWRAPPED_TOKEN_ACCOUNT> <ESCROW_ACCOUNT> <WRAPPED_TOKEN_PROGRAM> <RECIPIENT_TOKEN_ACCOUNT> <AMOUNT>

cargo run --bin spl-token-wrap wrap TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA 8eZ2rk3GcKopnKrAmVibV7zLW1sszNo2GdbhtxRhrZRi EYUCh7xP4sxnNBr1VFcGXfHm1piEA6ERci5prNykSV4G 27QMHK3W6qYni1abKz7tK7A6Sm69jvLaBv8zataRppEy TokenzQdBNbLqP5VEhdkAS6EPFLC1PHnBqCXEpPxuEb 6v5uYPBuubQqDwFU2uT6uQYC4P89H8iHzJgA4q7pvPHR 55

```

9. See console output
```bash
Wrapping 55 tokens from mint 8eZ2rk3GcKopnKrAmVibV7zLW1sszNo2GdbhtxRhrZRi
Unwrapped mint address: 8eZ2rk3GcKopnKrAmVibV7zLW1sszNo2GdbhtxRhrZRi
Wrapped mint address: 3xnGREX1Mi7zcuXJAYoDGk9xjpugZnfRhAKLPHdfEPPq
Unwrapped token account: EYUCh7xP4sxnNBr1VFcGXfHm1piEA6ERci5prNykSV4G
Recipient wrapped token account: 6v5uYPBuubQqDwFU2uT6uQYC4P89H8iHzJgA4q7pvPHR
Escrow account: 27QMHK3W6qYni1abKz7tK7A6Sm69jvLaBv8zataRppEy
Amount: 55
Signature: FRbCJmutYYFiGtnXzMFSgHucMRpdkHBHh4SCPX984qVQZRTfsFkvyKnnkbjW7M1nw1ZAKiTBC1igMnRSGg1H4x4
```

### CODE REVIEW

File: clients/cli/tests/helpers.rs [hunk 1]
```diff
@@ -75,36 +62,237 @@ pub async fn setup() -> Env {
...
-    let lamports = Rent::default().minimum_balance(data.len());
+    let blockhash = rpc_client.get_latest_blockhash().await.unwrap();
 
-    let account = Account {
-        lamports,
-        data,
-        owner: UNWRAPPED_TOKEN_PROGRAM,
-        ..Default::default()
-    };
-    (Pubkey::new_unique(), account.into())
```
Comment (grod220) [inline clients/cli/tests/helpers.rs:109]: Taking action on some feedback in last PR: https://github.com/solana-program/token-wrap/pull/33#pullrequestreview-2707626189

File: clients/cli/src/create_mint.rs [hunk 2]
```diff
@@ -50,6 +50,8 @@ pub struct CreateMintOutput {
     #[serde_as(as = "DisplayFromStr")]
     pub wrapped_mint_address: Pubkey,
     #[serde_as(as = "DisplayFromStr")]
+    pub wrapped_mint_authority: Pubkey,
```
[the remark this answers is not in the document]
Comment (buffalojoec) [reply clients/cli/src/create_mint.rs:53]: Do you need this in production, or only tests? Because it looks to me like you're deriving it.

File: clients/cli/src/wrap.rs [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+    /// The address of the token program that the wrapped mint should belong to
+    #[clap(value_parser = parse_token_program)]
+    pub wrapped_token_program: Pubkey,
+
+    /// The address of the token account to receive wrapped tokens
+    #[clap(value_parser = parse_pubkey)]
+    pub recipient_token_account: Pubkey,
+
+    /// The amount of tokens to wrap
+    #[clap(value_parser = |s: &str| -> Result<u64, String> { s.parse::<u64>().map_err(|e| e.to_string()) })]
```
Comment (buffalojoec) [inline clients/cli/src/wrap.rs:52]: You shouldn't need to be this explicit, right?

```suggestion
    #[clap(value_parser)]
```

File: clients/cli/tests/helpers.rs [hunk 4]
```diff
@@ -75,36 +62,237 @@ pub async fn setup() -> Env {
...
-    let state = spl_token::state::Mint {
-        decimals: 8,
-        is_initialized: true,
-        supply: 1_000_000_000,
-        mint_authority: COption::Some(*mint_authority),
-        freeze_authority: COption::None,
-    };
-    let mut data = vec![0u8; spl_token::state::Mint::LEN];
-    state.pack_into_slice(&mut data);
+pub async fn create_unwrapped_mint(
```
Comment (buffalojoec) [inline clients/cli/tests/helpers.rs:75]: For what it's worth, I prefer the startup account fixtures approach you had here originally. I saw the back-and-forth [over here](https://github.com/solana-program/token-wrap/pull/33#pullrequestreview-2707626189), so I'll leave it up to you to choose which you prefer, and it seems you have.

**Review by buffalojoec — COMMENTED** [review]
Looking great, just a few questions from my side, really.

File: clients/cli/src/wrap.rs [hunk 5]
```diff
@@ -0,0 +1,196 @@
...
+    #[clap(value_parser = parse_pubkey)]
+    pub escrow_account: Pubkey,
+
+    /// The address of the token program that the wrapped mint should belong to
+    #[clap(value_parser = parse_token_program)]
+    pub wrapped_token_program: Pubkey,
+
+    /// The address of the token account to receive wrapped tokens
+    #[clap(value_parser = parse_pubkey)]
+    pub recipient_token_account: Pubkey,
```
Comment (joncinque) [inline clients/cli/src/wrap.rs:49]: This could also be optional, and default to the ATA for the fee payer

**Review by joncinque — COMMENTED** [review]
Looks great overall! Mostly small things

File: clients/cli/tests/helpers.rs (hunk 4 above)
Comment (buffalojoec) [reply clients/cli/tests/helpers.rs:75]: > Do you have a reason for that preference?

I have formed a bias toward fixtures for a number of reasons, but it spawned from unit testing the RPC API for `@solana/kit` and has been hardened through writing program unit tests with Mollusk. It was also the motivation for building the latter.

The crux of the fixtures/mocks approach is the belief that your unit tests should be testing a single unit in isolation. If you're trying to test function C and its setup depends on functions A and B, now your tests for C are all coupled to functions A and B.

What if, all this time, `CreateMint` was doing something slightly erroneous, and we were getting _all_ false positives in `Wrap` and `Unwrap`? It might sound ridiculous, but it happens.

Fixtures/mocks also let you set up a wide array of edge cases, which might be complex to set up through transactions (perhaps they would require a custom program, for example).

I also think if we're going to write unit tests that use separate test-validator instances, the faster they spin up and then back down, the better. That means less "send and confirm".

> We can certainly stick with fixtures for unit tests, but for a CLI, an end-to-end test is more useful to test the whole flow, but maybe I'm missing some context.

I suppose I would just set up the per-command unit tests in isolation and then have one or more aptly-labeled end-to-end tests for the whole flow. I wouldn't make every CLI test an end-to-end test.

> We've had false positives in the past due to improper setup in tests, so I lean towards the more complicated setup to be sure we're actually testing a real-world situation.

Without knowing the context of these false positives I don't want to make any assumptions, but I generally think this is an acceptable risk compared to coupling too many things within each test.

Anyway, that's just my two cents! If you both disagree with me, you can do what you think is best!

Comment (joncinque) [reply clients/cli/tests/helpers.rs:75]: In my experience, there's a much higher chance of false positives with unit-testing than end-to-end testing because you can setup any set of inputs. As one example, remember all of the firedancer conformance fixtures that were checking impossible situations?

On the flip side, I'll definitely admit that highly-coupled tests are a lot more work to maintain. But I still stick by these words: [Write tests. Not too many. Mostly integration](https://kentcdodds.com/blog/write-tests)

For now, since we don't have integration tests otherwise, I lean towards doing them here in the CLI. But we can let Gabe be the tie-breaker :sweat_smile:

File: clients/cli/src/wrap.rs (hunk 5 above)
Comment (grod220) [reply clients/cli/src/wrap.rs:49]: Added this functionality and a test for it.

File: clients/cli/tests/helpers.rs (hunk 4 above)
Comment (grod220) [reply clients/cli/tests/helpers.rs:75]: Definitely can see the two sides to this. One benefit I've seen with these CLI tests making the actual calls is that it's shown me the real-life process the end-user will need to go through in order to make use of it. This inspired https://github.com/solana-program/token-wrap/pull/40 because I ran into a user flow I wasn't fully supporting.

One thing to note is in the main program code with mollusk, the processor methods were tested in isolation with pre-made accounts passed in, which feels like it worked well given the variety of cases those tests covered. Given that, I think this program probably could use the "integration coverage" by continuing along the path of this PR. Will go that route for these CLI tests.

File: clients/cli/Cargo.toml [hunk 6]
```diff
@@ -28,6 +27,7 @@ solana-signature = { workspace = true }
 solana-signer = { workspace = true }
 solana-system-interface = { workspace = true }
 solana-transaction = { workspace = true }
+spl-associated-token-account = { workspace = true }
```
Comment (joncinque) [inline clients/cli/Cargo.toml:30]: nit: can be down in follow-up work but `spl-associated-token-account-client` https://crates.io/crates/spl-associated-token-account-client/2.0.0 will give you what you need with fewer deps

(also, doesn't matter since the dep will change, but if using `spl-associated-token-account`, the `no-entrypoint` feature should also be enabled)

File: clients/cli/src/wrap.rs [hunk 7]
```diff
@@ -0,0 +1,224 @@
...
+    #[clap(value_parser)]
+    pub amount: u64,
+
+    /// Path to the signer for the transfer authority if different from
+    /// fee payer
+    #[clap(long = "transfer-authority", value_name = "PATH")]
+    pub transfer_authority: Option<String>,
+
+    /// The address of the mint to wrap, queried if not provided
+    #[clap(long = "unwrapped-mint", value_parser = parse_pubkey)]
```
Comment (joncinque) [inline clients/cli/src/wrap.rs:56]: nit: can remove the name here, since clap figures it out automatically
```suggestion
    #[clap(long, value_parser = parse_pubkey)]
```

File: clients/cli/src/wrap.rs [hunk 8]
```diff
@@ -0,0 +1,224 @@
...
+    pub unwrapped_mint: Option<Pubkey>,
+
+    /// The address of the token account to receive wrapped tokens.
+    /// If not provided, defaults to fee payer associated token account
+    #[clap(long = "recipient-token-account", value_parser = parse_pubkey)]
```
Comment (joncinque) [inline clients/cli/src/wrap.rs:61]: nit: same here, let clap figure it out automatically
```suggestion
    #[clap(long, value_parser = parse_pubkey)]
```

File: clients/cli/src/wrap.rs [hunk 9]
```diff
@@ -0,0 +1,224 @@
...
+    #[clap(value_parser = parse_token_program)]
+    pub wrapped_token_program: Pubkey,
+
+    /// The amount of tokens to wrap
+    #[clap(value_parser)]
+    pub amount: u64,
+
+    /// Path to the signer for the transfer authority if different from
+    /// fee payer
+    #[clap(long = "transfer-authority", value_name = "PATH")]
```
Comment (joncinque) [inline clients/cli/src/wrap.rs:52]: nit: same here, let clap figure it out automatically
```suggestion
    #[clap(long, value_name = "PATH")]
```

File: clients/cli/src/wrap.rs [hunk 10]
```diff
@@ -0,0 +1,224 @@
...
+
+impl QuietDisplay for WrapOutput {
+    fn write_str(&self, _: &mut dyn std::fmt::Write) -> std::fmt::Result {
+        Ok(())
+    }
+}
+impl VerboseDisplay for WrapOutput {}
+
+async fn get_unwrapped_mint(
+    rpc_client: &Arc<RpcClient>,
```
Comment (joncinque) [inline clients/cli/src/wrap.rs:123]: nit: passing in a `&Arc<T>` is an antipattern if the Arc isn't needed, it's simpler and clearer to just take in the ref directly
```suggestion
    rpc_client: &RpcClient,
```

File: clients/cli/src/wrap.rs [hunk 11]
```diff
@@ -0,0 +1,224 @@
...
+        rc::Rc,
+        sync::Arc,
+    },
+};
+
+#[derive(Clone, Debug, Args)]
+pub struct WrapArgs {
+    /// The address of the token program that the unwrapped mint belongs to
+    #[clap(value_parser = parse_token_program)]
+    pub unwrapped_token_program: Pubkey,
```
Comment (joncinque) [inline clients/cli/src/wrap.rs:32]: This can be optional as well, inferred from the unwrapped token account

File: clients/cli/tests/helpers.rs [hunk 12]
```diff
@@ -75,36 +62,227 @@ pub async fn setup() -> Env {
...
+        )
+        .unwrap()
+    } else {
+        spl_token_2022::instruction::initialize_account(
+            token_program,
+            &token_account.pubkey(),
+            mint,
+            owner,
+        )
+        .unwrap()
```
Comment (joncinque) [inline clients/cli/tests/helpers.rs:234]: nit: the spl_token_2022 constructor works for both programs, so simpler just to use that
```suggestion
    let initialize_instruction = spl_token_2022::instruction::initialize_account(
            token_program,
            &token_account.pubkey(),
            mint,
            owner,
        )
        .unwrap()
```

**Review by joncinque — COMMENTED** [review]
Just a few last nits, then this can go in from my side

File: clients/cli/Cargo.toml (hunk 6 above)
Comment (grod220) [reply clients/cli/Cargo.toml:30]: Simple enough to update that dep 👍

**Review by joncinque — APPROVED** [review]
Looks great!

**Review by buffalojoec — APPROVED** [review]
Ship it!
